### PR TITLE
Pictures: Move image metadata processing to exiv2 as a new dependency

### PR DESCRIPTION
**Pull request by enen92**
## Description
This provides a revamp of our picture metadata (exif and iptc) extraction handling logic moving it from a custom implementation (untouched for 14 years - and rencently broken (and fixed) for most of the tags) to an external dependency (exiv2). This allows us to use a single dependency for all image metadata needs and also avoid a lot of corner cases we have currently in our own handling logic while contributing directly to the open-source ecosystem.

**Upstream contributions:**
Add EXIV2_ENABLE_FILESYSTEM_ACCESS option - https://github.com/Exiv2/exiv2/pull/2837
Set conditional HTTP depending on EXIV2_ENABLE_WEBREADY - https://github.com/Exiv2/exiv2/pull/2844
jpeg: add encodingProcess and numColorComponents SOF members - https://github.com/Exiv2/exiv2/pull/2874
Disable psapi in UWP (unsupported): https://github.com/Exiv2/exiv2/pull/2990

## Motivation and context
Custom logic we have in Kodi without any maintainers should be delegated as much as possible to well maintained libs. Exiv2 is a good example, well maintained, easy to contribute and is part of a lot of software projects (e.g. gimp). Also cleans up a lot of magic in GUIPictureInfo.

## How has this been tested?
Runtime tested in windows, android, macOS as well as UWP.

## What is the effect on users?
None, this is a 1:1 match of our current implementation. I'll track eventual regressions.

## Screenshots (if appropriate):

IPTC:
image

EXIF:
image

image

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [x] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [x] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [x] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

File: tools/depends/target/Makefile [hunk 1]
```diff
@@ -186,6 +187,7 @@ taglib: $(ZLIB)
 wayland: expat libffi
 waylandpp: wayland $(WAYLANDPP_DEPS)
 xz: gettext
+exiv2: $(ICONV) $(ZLIB) expat
```
Comment (garbear) [inline tools/depends/target/Makefile:190]: alphabetical order please

File: tools/depends/target/exiv2/Makefile [hunk 2]
```diff
@@ -0,0 +1,40 @@
+include ../../Makefile.include EXIV2-VERSION ../../download-files.include
+DEPS = ../../Makefile.include Makefile EXIV2-VERSION ../../download-files.include 0001-Remove-filesystem-dependency.patch
+
+# configuration settings
+CMAKE_OPTIONS=-DCMAKE_INSTALL_PREFIX=$(PREFIX) \
+			  -DEXIV2_ENABLE_WEBREADY=OFF \
```
Comment (garbear) [inline tools/depends/target/exiv2/Makefile:6]: Indentation here is a mix of tabs and spaces, maybe remove the 3 tabs and just use 2 spaces?

**Comment by garbear** [comment]
great work so far! looking forward to this reaching completion for P*.

**Review by neo1973 — COMMENTED** [review]
Looks good so far, just some suggestion for things I noticed when browsing the code :+1:

File: xbmc/pictures/metadata/ImageMetadataParser.cpp [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+  auto image = Exiv2::ImageFactory::open(outputBuffer.data(), readbytes);
+  image->readMetadata();
+
+  // extract metadata
+  ExtractCommonMetadata(image);
+
+  ExtractExif(image->exifData());
+  ExtractIPTC(image->iptcData());
+
+  return std::move(m_imageMetadata);
```
[the remark this answers is not in the document]
Comment (enen92) [reply xbmc/pictures/metadata/ImageMetadataParser.cpp:88]: @neo1973 I get `Call to implicitly-deleted copy constructor of 'std::unique_ptr<ImageMetadata>'`

[the remark this answers is not in the document]
Comment (neo1973) [reply xbmc/pictures/metadata/ImageMetadataParser.cpp:81]: Right. Somehow I overlooked the fact that this is a member variable :slightly_smiling_face:

File: xbmc/pictures/metadata/ImageMetadataParser.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+#include <exiv2/exiv2.hpp>
+
+class CImageMetadataParser
+{
+public:
+  CImageMetadataParser();
+  ~CImageMetadataParser() = default;
+
+  std::unique_ptr<ImageMetadata> ExtractMetadata(const std::string& picFileName);
```
Comment (neo1973) [inline xbmc/pictures/metadata/ImageMetadataParser.h:24]: How about making this method static? There seems no point in instantiating a `CImageMetadataParser`, right?

<details>
<summary>Proposed change</summary>

```diff
diff --git a/xbmc/pictures/PictureInfoTag.cpp b/xbmc/pictures/PictureInfoTag.cpp
index 02857a2ddd..7c63f5db72 100644
--- a/xbmc/pictures/PictureInfoTag.cpp
+++ b/xbmc/pictures/PictureInfoTag.cpp
@@ -62,8 +62,7 @@ bool CPictureInfoTag::Load(const std::string &path)
   // Load by Kodi's included own way
   if (!m_isLoaded)
   {
-    CImageMetadataParser metadataParser;
-    std::unique_ptr<ImageMetadata> metadata = metadataParser.ExtractMetadata(path);
+    std::unique_ptr<ImageMetadata> metadata = CImageMetadataParser::ExtractMetadata(path);
     if (metadata)
     {
       m_imageMetadata = *(metadata.release());
diff --git a/xbmc/pictures/metadata/ImageMetadataParser.cpp b/xbmc/pictures/metadata/ImageMetadataParser.cpp
index e991b642d5..a2c14c49d4 100644
--- a/xbmc/pictures/metadata/ImageMetadataParser.cpp
+++ b/xbmc/pictures/metadata/ImageMetadataParser.cpp
@@ -72,13 +72,15 @@ std::unique_ptr<ImageMetadata> CImageMetadataParser::ExtractMetadata(const std::
   auto image = Exiv2::ImageFactory::open(outputBuffer.data(), readbytes);
   image->readMetadata();
 
+  CImageMetadataParser parser;
+
   // extract metadata
-  ExtractCommonMetadata(image);
+  parser.ExtractCommonMetadata(image);
 
-  ExtractExif(image->exifData());
-  ExtractIPTC(image->iptcData());
+  parser.ExtractExif(image->exifData());
+  parser.ExtractIPTC(image->iptcData());
 
-  return std::move(m_imageMetadata);
+  return std::move(parser.m_imageMetadata);
 }
 
 void CImageMetadataParser::ExtractCommonMetadata(std::unique_ptr<Exiv2::Image>& image)
diff --git a/xbmc/pictures/metadata/ImageMetadataParser.h b/xbmc/pictures/metadata/ImageMetadataParser.h
index b7fa623da1..8117c30a02 100644
--- a/xbmc/pictures/metadata/ImageMetadataParser.h
+++ b/xbmc/pictures/metadata/ImageMetadataParser.h
@@ -18,12 +18,13 @@
 class CImageMetadataParser
 {
 public:
-  CImageMetadataParser();
   ~CImageMetadataParser() = default;
 
-  std::unique_ptr<ImageMetadata> ExtractMetadata(const std::string& picFileName);
+  static std::unique_ptr<ImageMetadata> ExtractMetadata(const std::string& picFileName);
 
 private:
+  CImageMetadataParser();
+
   void ExtractCommonMetadata(std::unique_ptr<Exiv2::Image>& image);
   void ExtractExif(Exiv2::ExifData& exifData);
   void ExtractIPTC(Exiv2::IptcData& iptcData);
diff --git a/xbmc/pictures/metadata/test/TestMetadataExtraction.cpp b/xbmc/pictures/metadata/test/TestMetadataExtraction.cpp
index ff61ecc881..96852efa58 100644
--- a/xbmc/pictures/metadata/test/TestMetadataExtraction.cpp
+++ b/xbmc/pictures/metadata/test/TestMetadataExtraction.cpp
@@ -13,16 +13,6 @@
 
 #include <gtest/gtest.h>
 
-namespace
-{
-std::unique_ptr<ImageMetadata> ExtractMetadata(const std::string& path)
-{
-  CImageMetadataParser metadataParser;
-  return metadataParser.ExtractMetadata(path);
-}
-
-} // namespace
-
 class TestMetadataExtraction : public ::testing::Test
 {
 protected:
@@ -32,7 +22,7 @@ protected:
 TEST_F(TestMetadataExtraction, TestGPSImage)
 {
   auto path = XBMC_REF_FILE_PATH("xbmc/pictures/metadata/test/testdata/exifgps.jpg");
-  std::unique_ptr<ImageMetadata> metadata = ExtractMetadata(path);
+  std::unique_ptr<ImageMetadata> metadata = CImageMetadataParser::ExtractMetadata(path);
   // Extract GPS tags
   EXPECT_EQ(metadata->exifInfo.GpsLat, "49° 5' 47.469\" N");
   EXPECT_EQ(metadata->exifInfo.GpsLong, "9° 23' 14.630\" E");
@@ -63,7 +53,7 @@ TEST_F(TestMetadataExtraction, TestGPSImage)
 TEST_F(TestMetadataExtraction, TestIPTC)
 {
   auto path = XBMC_REF_FILE_PATH("xbmc/pictures/metadata/test/testdata/iptc.jpg");
-  std::unique_ptr<ImageMetadata> metadata = ExtractMetadata(path);
+  std::unique_ptr<ImageMetadata> metadata = CImageMetadataParser::ExtractMetadata(path);
   // Check the IPTC tags of this image
   EXPECT_TRUE(metadata->iptcInfo.Keywords.empty());
   EXPECT_EQ(metadata->iptcInfo.RecordVersion, "4");
```
</details>

Comment (enen92) [reply xbmc/pictures/metadata/ImageMetadataParser.h:24]: Yes makes total sense, thanks!

**Comment by enen92** [comment]
jenkins build this please

**Comment by enen92** [comment]
Finally found some time to rebase and finish this. Decided to remove all the new extra features (new infolabels) and keep it a 1:1 version of our current implementation.
Hopefully someone can take a look.

**Comment by enen92** [comment]
Finally found some time to rebase and finish this. Decided to remove all the new extra features (new infolabels) and keep it a 1:1 version of our current implementation.
Hopefully someone can take a look.

File: cmake/modules/FindIconv.cmake [hunk 5]
```diff
@@ -31,14 +31,11 @@ if(NOT TARGET ${APP_NAME_LC}::${CMAKE_FIND_PACKAGE_NAME})
...
-                                                                       INTERFACE_INCLUDE_DIRECTORIES "${ICONV_INCLUDE_DIR}")
-    endif()
-  else()
-    if(Iconv_FIND_REQUIRED)
-      message(FATAL_ERROR "Iconv libraries were not found.")
+      add_library(Iconv::Iconv UNKNOWN IMPORTED)
+      set_target_properties(Iconv::Iconv PROPERTIES
+                                         IMPORTED_LOCATION "${ICONV_LIBRARY}"
+                                         INTERFACE_INCLUDE_DIRECTORIES "${ICONV_INCLUDE_DIR}")
+      set_property(GLOBAL APPEND PROPERTY INTERNAL_DEPS_PROP Iconv::Iconv)
```
Comment (fuzzard) [inline cmake/modules/FindIconv.cmake:38]: Don't go changing this back. This needs another solution

Comment (enen92) [reply cmake/modules/FindIconv.cmake:38]: Can you please let me know how to handle it? I just cherry picked your change from months ago. Cmake is black magic to me

Comment (fuzzard) [reply cmake/modules/FindIconv.cmake:38]: Yeah will do when I get some time to think about it. Will try to not take too long

Comment (fuzzard) [reply cmake/modules/FindIconv.cmake:38]: https://github.com/fuzzard/xbmc/commit/27924f4082b0a64e2d75766c7cddd95c7df8930d for the standardisation of the FindExiv.cmake file

https://github.com/fuzzard/xbmc/commit/3fd59dff8690cff660e29adcbe24e76179fd8c28 for the iconv requirements. just put it as one of the first commits in the PR. Alternatively i can PR this separately, but the approval turn around will probably be slow causing this PR to be held up.

File: tools/depends/target/Makefile (hunk 1 above)
Comment (fuzzard) [reply tools/depends/target/Makefile:191]: Still a rebase issue here. Its correct in "[depends] Add exiv2 to depends" commit, then this line is added back in "Pictures: Move to exiv2 for image metadata retrieval (exif and iptc)"

**Comment by fuzzard** [comment]
apart from the cmake stuff, everything else looked good. so thumbs up from me

File: tools/depends/target/Makefile (hunk 1 above)
Comment (enen92) [reply tools/depends/target/Makefile]: fixed, thanks

File: tools/depends/target/exiv2/EXIV2-VERSION [hunk 6]
```diff
@@ -1,6 +1,7 @@
 LIBNAME=exiv2
 VERSION=0.28.2
-ARCHIVE=$(LIBNAME)-$(VERSION).tar.gz
+BASE_URL=https://github.com/Exiv2/exiv2/archive/refs/tags/
```
Comment (fuzzard) [inline tools/depends/target/exiv2/EXIV2-VERSION:3]: This commit can be removed when you are ready. File is on mirrors

File: cmake/modules/FindExiv2.cmake [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+      add_library(EXIV2::EXIV2 ALIAS Exiv2::exiv2lib)
+    else()
+      add_library(EXIV2::EXIV2 UNKNOWN IMPORTED)
+      set_target_properties(EXIV2::EXIV2 PROPERTIES
+                                         IMPORTED_LOCATION "${EXIV2_LIBRARY}"
+                                         INTERFACE_INCLUDE_DIRECTORIES "${EXIV2_INCLUDE_DIR}")
+      if(CORE_SYSTEM_NAME STREQUAL "freebsd")
+        set_property(TARGET EXIV2::EXIV2 APPEND PROPERTY
+                                                INTERFACE_LINK_LIBRARIES procstat)
+      elseif(CORE_SYSTEM_NAME MATCHES "windows")
```
[the remark this answers is not in the document]
Comment (enen92) [reply cmake/modules/FindExiv2.cmake:120]: I've fixed this upstream in https://github.com/Exiv2/exiv2/pull/2990

**Comment by enen92** [comment]
All patches included have been submitted and merged upstream. Three of them where already backported to 0.23.x, just one pending but should be close (https://github.com/Exiv2/exiv2/pull/2994). This means we'll probably bump exiv2 with zero patches in a few weeks/months.

I plan to merge this by the end of the week (sunday night) so If you have any concerns please voice over.

**Comment by fuzzard** [comment]
Test failure completely unrelated, and build failure from previous 2 jobs were builder related and not a test failure. Merging now

**Comment by thexai** [comment]
exiv2 fails to build in debug mode:

`cl : command line error D8016: '/Ox' and '/RTC1' command-line options are incompatible`

to reproduce:

```
cmake -G "Visual Studio 17 2022" -A x64 -T host=x64 V:\kodi

cmake --build . --config "Debug" --target exiv2
```

Seems is due this:
https://github.com/Exiv2/exiv2/blob/e843bc4834aff138db7dc87bd2f4dfd98c49de4a/cmake/compilerFlags.cmake#L136

**Comment by enen92** [comment]
@fuzzard I guess I need your help. Any way of guarding that set call if RTC1 is enabled? For what I can tell it seems to be set by cmake build type debug automatically.

**Comment by fuzzard** [comment]
There isnt. your options are either patch it out of the exiv lib, or alter our flags we pass through. We do option 2 for the effects11 library already for a pointer to how to do it on our side

https://github.com/xbmc/xbmc/blob/5069e47861783cd78e3c758f3212e6147b817aaf/cmake/modules/FindEffects11.cmake#L17-L26

**Comment by mkreisl** [comment]
Sorry, something wrong finding the files:

```
-- EXIV2_URL: https://mirrors.kodi.tv/build-deps/sources/exiv2-0.28.2.tar.gz
CMake Error at /usr/share/cmake-3.25/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find Exiv2 (missing: EXIV2_LIBRARY) (found version "0.27.6")
Call Stack (most recent call first):
  /usr/share/cmake-3.25/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  cmake/modules/FindExiv2.cmake:104 (find_package_handle_standard_args)
  cmake/scripts/common/Macros.cmake:390 (find_package)
  cmake/scripts/common/Macros.cmake:404 (find_package_with_ver)
  CMakeLists.txt:269 (core_require_dep)
```

dev package(s) are installed:

```
dpkg -l | grep exiv
ii  libexiv2-27:arm64                     0.27.6-1                            arm64        EXIF/IPTC/XMP metadata manipulation library
ii  libexiv2-27:armhf                     0.27.6-1                            armhf        EXIF/IPTC/XMP metadata manipulation library
ii  libexiv2-dev:arm64                    0.27.6-1                            arm64        EXIF/IPTC/XMP metadata manipulation library - development files
ii  libexiv2-dev:armhf                    0.27.6-1                            armhf        EXIF/IPTC/XMP metadata manipulation library - development files
```

**Comment by enen92** [comment]
@mkreisl version 0.28.2 is required (you seem to have 0.27.6):
https://github.com/xbmc/xbmc/blob/master/tools/depends/target/exiv2/EXIV2-VERSION#L2

**Comment by mkreisl** [comment]
WTF, that's what Debian Bookworm has

The 0.28 version is not even available in Sid. I don't think that makes sense

**Comment by enen92** [comment]
I was about to recommend to use `ENABLE_INTERNAL_EXIV2`.

**Comment by mkreisl** [comment]
> I was about to recommend to use `ENABLE_INTERNAL_EXIV2`.

Ah yes, I somehow lost sight of that, would be a cool option if the external version doesn't meet the requirements to build the internal one. But anyway. the way it was solved here now works for me too.

**Pull request merged**